test/app-luatest/gh_3680_export_all_luajit_symbols_test.lua

```diff
@@ -0,0 +1,278 @@
+    'luaL_error',
+    'luaL_getmetafield',
+    -- <luaL_getmetatable> is a macro to <lua_getfield> + LUA_REGISTRYINDEX.
+    'luaL_gsub',
+    'luaL_loadbuffer',
+    'luaL_loadfile',
+    'luaL_loadstring',
+    'luaL_newmetatable',
+    'luaL_newstate',
+    'luaL_openlibs',
+    -- <luaL_optint> is a macro to <luaL_optinteger>.
+    'luaL_optinteger',
+    -- <luaL_optlong> is a macro to <luaL_optinteger>.
+    'luaL_optlstring',
+    'luaL_optnumber',
+    -- <luaL_optstring> is a macro to <luaL_optlstring>.
+    'luaL_prepbuffer',
+    'luaL_pushresult',
+    'luaL_ref',
+    'luaL_register',
+    -- <luaL_typename> is a macro to <lua_typename>.
+    'luaL_typerror',
+    'luaL_unref',
+    'luaL_where',
+
+    -- Interfaces from Lua 5.2 Reference Manual (i.e. Lua C API
+    -- functions from lua.h).
+    -- XXX: Interfaces implemented via C macros are also listed
+    -- below only for consistency.
+    'lua_copy',
+    'lua_loadx',
+    'lua_tointegerx',
+    'lua_tonumberx',
+    'lua_upvalueid',
+    'lua_upvaluejoin',
+    'lua_version',
+
+    -- Auxiliary interfaces from Lua 5.2 Reference Manual (i.e.
+    -- Lua auxiliary library functions from lauxlib.h).
+    -- XXX: Interfaces implemented via C macros are also listed
+    -- below only for consistency.
+    'luaL_execresult',
+    'luaL_fileresult',
+    'luaL_loadbufferx',
+    'luaL_loadfilex',
+    -- <luaL_newlibtable> is a macro to <lua_createtable>.
+    -- <luaL_newlib> is a macro to <luaL_newlibtable> + <luaL_setfuncs>.
+    'luaL_setfuncs',
+    'luaL_setmetatable',
+    'luaL_testudata',
+    'luaL_traceback',
+
+    -- Interfaces from Lua 5.3 Reference Manual (i.e. Lua C API
+    -- functions from lua.h).
+    -- XXX: Interfaces implemented via C macros are also listed
+    -- below only for consistency.
+    'lua_isyieldable',
+    -- Auxiliary interfaces from Lua 5.3 Reference Manual (i.e.
+    -- Lua auxiliary library functions from lauxlib.h).
+    -- XXX: Interfaces implemented via C macros are also listed
+    -- below only for consistency.
+    -- <luaL_opt> is a macro to <lua_isnoneornil>.
+
+    -- Auxiliary interfaces provided by LuaJIT (i.e. public
+    -- functions from lauxlib.h, which are not listed in any Lua
+    -- Reference Manual).
+    'luaL_findtable',
+    'luaL_openlib',
+    'luaL_pushmodule',
+
+    -- Interfaces provided by LuaJIT (i.e. public functions from
+    -- luajit.h).
+    'luaJIT_profile_dumpstack',
+    'luaJIT_profile_start',
+    'luaJIT_profile_stop',
+    'luaJIT_setmode',
+    'luaJIT_version_2_1_0_beta3',
+
+    -- Interfaces provided by Tarantool LuaJIT fork.
+    'luaM_metrics',
+    'luaM_sysprof_report',
+    'luaM_sysprof_set_backtracer',
+    'luaM_sysprof_set_on_stop',
+    'luaM_sysprof_set_writer',
+    'luaM_sysprof_start',
+    'luaM_sysprof_stop',
+}
+
+local hidden = {
+    -- Interfaces to load particular Lua or LuaJIT builtin module.
+    -- All of them are hidden in favor to <lua_openlibs> function.
```

Comment:
Well, I've expressed my opinion via code and comment. I have no strict objections to your suggestion regarding `luaopen_*` interfaces, but I would like to know @Totktonada opinion before making any changes.

Reply:
I would stay on the compatibility side here.

My thoughts here are the following.

* These calls are in the Lua/LuaJIT public API and I guess they're no-op on an initialized lua_State. Can be used in some common code with an idea that if it is redundant, it is no-op.
* In theory a module may create some completely clean state (using `luaL_newstate()`) for some, hmm, isolation/sandboxing... multithreading... specific resource cleanup flow? And use the deprecated API, because... this code is old?
* It is part of a language runtime library, not just some library.
* Tarantool is not the most well known implementation of the language, so we can't make a decision to deprecate some Lua API and expect that the community will adopt it. IOW, an incompatibility would hit a wrong person -- not an author of the deprecated code, but its (possibly indirect) user.
* We don't have to pay much maintanance effort to keep the functions. We can move them into some 'APIs from Lua 5.0 deprecated in Lua 5.1' group.

OTOH, I guess that removing the functions unlikely will have any visible consequences for our users. So, I don't insist here and would agree with Igor's decision.

Reply:
OK, we've already discussed this a couple of times with @Buristan and considering @Totktonada inputs, I agree to leave it in the separate section for Lua 5.0 legacy API (I found some kind of listing in the contents of [this PDF](https://www.lua.org/ftp/refman-5.0.pdf)). It would be great to enable some kind of warnings for using such API, but definitely not in the scope of this changeset.